survey_mailer/survey_mailer.go

```diff
@@ -0,0 +1,204 @@
+
+		log.Println("Sending chapter meeting survey for event:", event.EventName)
+
+		// send all emails, including "missing" email
+		bulkSendEmails(event, subject, bodyText, bodyHtml)
+
+		// updateSurveyStatus
+		updateSurveyStatus(db, event.ID)
+	}
+}
+
+func surveyProtest(db *sqlx.DB, queryDate string) {
+
+	log.Println("Looking for protest events on", queryDate)
+
+	// Get yesterday's protest events that haven't had surveys sent yet
+	events, err := model.GetEvents(db, model.GetEventOptions{
+		DateFrom:   queryDate,
+		DateTo:     queryDate,
+		EventType:  "%action",
+		SurveySent: "0",
+	})
+
+	if err != nil {
+		log.Printf("Failed to get today's protest events: %v", err)
+		return
+	}
+
+	// Iterate through events
+	for _, event := range events {
+
+		// Set email subject & body
+		subject := "Survey: " + event.EventName
+		eventNameParam := strings.Replace(event.EventName, " ", "+", -1)
+		bodyText := `Thank you for taking part in direct action! Please take this quick survey: https://docs.google.com/forms/d/e/1FAIpQLScfrPtPxmYAroODhBkwUGq753JPykYKNdosg4gUR_SRng8BRQ/viewform?usp=pp_url&entry.466557185=` + eventNameParam + ". If you captured any photos or videos, please upload them here: dxe.io/upload."
+		bodyHtml := `<p>Thank you for taking part in direct action! Please <a href="https://docs.google.com/forms/d/e/1FAIpQLScfrPtPxmYAroODhBkwUGq753JPykYKNdosg4gUR_SRng8BRQ/viewform?usp=pp_url&entry.466557185=` + eventNameParam + `">click here</a> to take a quick survey.</p><p>If you captured any photos or videos, please upload them <a href="http://dxe.io/upload">here</a>.</p>`
+
+		log.Println("Sending protest survey for event:", event.EventName)
+
+		// send all emails, including "missing" email
+		bulkSendEmails(event, subject, bodyText, bodyHtml)
+
+		// updateSurveyStatus
+		updateSurveyStatus(db, event.ID)
+	}
+}
+
+func surveyMailerWrapper(db *sqlx.DB) {
+	defer func() {
+		if r := recover(); r != nil {
+			log.Println("Recovered from panic in survey mailer", r)
+		}
+	}()
+
+	// Get current hour of day & current day of week
+	weekday := time.Now().Weekday()
+	hour := time.Now().Hour()
+	// Calculate date of yesterday & today
+	today := time.Now().Format("2006-01-02")
+	yesterday := time.Now().AddDate(0, 0, -1).Format("2006-01-02")
+
+	// send protest surveys from previous day daily at 9am
+	if hour == 9 {
+		surveyProtest(db, yesterday)
+	}
+
+	// send meetup surveys on saturdays at 2pm
+	if weekday == 6 && hour == 14 {
+		surveyMeetup(db, today)
+	}
+
+	// send chapter mtg surveys on sun at 4pm
+	if weekday == 0 && hour == 16 {
+		surveyChapterMeeting(db, today)
+	}
+
+}
+
+// Sends surveys based on event attendance every 60 minutes.
+// Should be run in a goroutine.
+func StartSurveyMailer(db *sqlx.DB) {
+	for {
+		log.Println("Starting survey mailer")
+		surveyMailerWrapper(db)
+		log.Println("Finished survey mailer")
+		// TODO(jake): change this to 60 min before deploying
+		time.Sleep(60 * time.Minute)
```

Comment:
I'd write this as:

    for now := range time.Tick(time.Hour) {
        ...
        surveyMailerWrapper(db, now)
        ...
    }

That ensures the clock doesn't drift over time.

I'm somewhat concerned there's still a chance that you get one event at 8:59:59.999999 and then because of timer delays, the next event doesn't happen until 10:00:00.000000 and you completely miss the events at 9am. I'll dig into the timer code to see if this is a risk.

Reply:
Do you think it might be better to have this run every minute to check what time it is, then only have the other functions run when the time is exactly (to the minute) 9:00am, 2pm, or 4pm?

Alternatively, we could just have it run every 30 minutes, then there are two chances for it to run if something goes wrong. (And duplicates wouldn't be sent because things get marked as "sent" in the database.)

Reply:
Thinking about it some more, it would probably be best to use a package like https://godoc.org/github.com/robfig/cron.

However, another thing to consider would be if there's a way to generalize the logic further, so you don't even have to do different logic at fixed schedules.

E.g., the real logic here seems to be that you want to process all events that haven't been processed yet, but you want to wait until a certain amount of time until after the event.

The former is already encoded in the database already. Could you check for the latter by looking for events where the time was at least N hours ago?

Reply:
Hmmm. Yes, it would work to just send surveys ~6 hours after events are created. Currently we aren't storing event time in the database though, so I'd have to add that first. And I'm slightly worried that doing so might create new issues.

An easier solution might be to just always send surveys the day after events. I don't think anyone would mind that.

Reply:
Ah, forgot we only have date, not time.

Reply:
> An easier solution might be to just always send surveys the day after events. I don't think anyone would mind that.

I use this implementation in the next commit.